### DISCUSSION TITLE
Authenticated access to content

### DISCUSSION THREAD

**Discussion by oripka** [nuxt/content#1532] [Q&A]
I'm using Nuxt Content as a tutorial / lab guide system for some courses. It's really nice to be able to do little quizes by implementing them as MDC.

I also implemented a way to convert the the markdown content to PDF, as students of courses often want to have a definitive PDF to take home with them (in the fear that the website will not be around forever). This is based upon pandoc, eisvogel.tex, some python scripting and Docker, CI/CD.

Back to my original question. I found this issue https://github.com/nuxt/content/issues/520 for v1. There is some talk about implementing authenticated access to `content` in v2. Is that actually possible now? It would be really nice to not have to put the markdown into a database in order to protect it.

Until now I figured that I can just use the authentication in the Vue Components but I am not sure how to properly prevent access to the markdown content.

All the best

**Comment by oripka** [comment]
I now solved this by overwriting `pages/document-driven.vue` from vue content, adding a middleware:

```
definePageMeta({
    middleware: 'auth'
})
```

The middleware is using supabase

```
export default defineNuxtRouteMiddleware((to, _from) => {
    const user = useSupabaseUser()
    if (!user.value) {
        return navigateTo('/')
    }
})
```

Additionally overwriting the navigation of Docus in `components/content/NavbarCenter.vue`

```
<script setup lang="ts">
[...]
const { hasPermissionToCourse, getCourseName } = useCourses();

const tree = computed(() => {
  return (navigation.value || []).filter((item) => {
    let hasPermission = hasPermissionToCourse(item.title)
    if (filtered.value.includes(item._path) || !hasPermission) { return false }
    return true
  })
})
[...]
</script>

<template>
  <nav v-if="hasNavigation"
    class="flex items-center justify-center flex-1 max-w-full space-x-2 overflow-hidden font-medium truncate">
    <template v-for="link in tree" :key="link._path">
      <NuxtLink :to="navBottomLink(link)" class="link"
        :class="{ active: isActive(link) }">
        <Icon v-if="link.icon && docus.header.showLinksIcons" :name="link.icon" class="w-4 h-4" />
        {{getCourseName(link.title)}}
      </NuxtLink>
    </template>
  </nav>
</template>
[...]

```

**Comment by oripka** [comment]
I just realized that this still leaves `server/api/query.ts` unprotected.

**Comment by sudomk** [comment]
Hi @oripka,
did you get any further with this 'problem'?

**Comment by oripka** [comment]
Additionally to what I mentioned above. I found out that one can protect `server/api/query.ts` by implementing a server middleware. 

But I also noted that it doesn't fully protect the content e.g. somebody can still just download `api/_content/cache.json` to get all the content.

I read through some of the Content and Docus code but my I still feel that I don't understand exactly how to do this with the least  amount of friction.

`middlware/auth.ts`

```

export default defineNuxtRouteMiddleware( async (to, _from) => {

    const { authentication } = useRuntimeConfig()

    if(!authentication.required){
        return 
    }
    
    const user = useSupabaseUser()
    const { hasPermissionToCoursePath, getCoursePermissions } = useCourses();
    
    if (!user.value) {
        return navigateTo('/')
    }

    await getCoursePermissions()
    
    if(to.meta.page !== false && !hasPermissionToCoursePath(to.fullPath) ){
        return navigateTo('/')
    }
})
```

`pages/document-driven.vue`

```
<script setup lang="ts">
import { useContent, useContentHead, useRequestEvent } from '#imports'

const { page, layout } = useContent()

// Page not found, set correct status code on SSR

if (!page.value && process.server) {
    const event = useRequestEvent()
    event.res.statusCode = 404
}

definePageMeta({
    middleware: 'auth'
})

useContentHead(page)
</script>
    
<template>
    <div class="document-driven-page">
        <NuxtLayout :name="layout || 'default'">
            <ContentRenderer v-if="page" :key="page._id" :value="page">
                <template #empty="{ value }">
                    <DocumentDrivenEmpty :value="value" />
                </template>
            </ContentRenderer>
            <DocumentDrivenNotFound v-else />
        </NuxtLayout>
    </div>
</template>
```

Here the server middleware (code is bit rough right now, but you should get the idea.)
`server/middleware/auth.ts`

```
import { serverSupabaseClient } from '#supabase/server'
import { serverSupabaseUser } from '#supabase/server'


function isActiveExercise(path, allowlist) {

    var isMatch = allowlist.some(
        function (rx) {
            return new RegExp(rx).test(path);
        }
    );
    return isMatch
}

export default defineEventHandler(async (event) => {
    return
    const query = useQuery(event)

    if (!event.req?.url) return

    let url = event.req?.url
    // let user = await serverSupabaseUser(event)

    if (query && url.includes("/api/_content/query")) {

        // console.log(user)
        // if (!user) {
        //     throw createError({
        //         statusCode: 403,
        //         statusMessage: 'Unauthorized'
        //     })
        // }

        const client = serverSupabaseClient(event)

        const { data: myCourses, error } = await client.from('user_courses').select(`courseid, show, courses (name)`)


        // const activeExercises = [
        //     "^/.*/intro",
        //     "^/.*/bsi",
        //     "^/.*/info",
        //     //"^.*" //everything
        //     //"^/.*/pads",
        //     //"^/cysecde/windows$",
        //     //"^/cysecde/windows/pivot-to-windows$"
        // ]
        let activeExercisesTmp = []
        for (let c of myCourses) {
            if (c.show) {
                activeExercisesTmp.push(...c.show)
            }
        }
        let activeExercises = [...new Set(activeExercisesTmp)]
        
        let p = JSON.parse(query._params)
        let path

        if (p.surround) {
            path = p.surround.query
        } else if (p.where) {
            path = p.where[0]._path
        }

        if (!isActiveExercise(path, activeExercises)) {
            throw createError({
                statusCode: 403,
                statusMessage: 'Unauthorized'
            })
        }
    }


})
```

With the server middleware I ran into performance problems and also the issue that `let user = await serverSupabaseUser(event)` is undefined if one directly browses to a certain nuxt content page (https://github.com/nuxt-modules/supabase/issues/104)

**Comment by onpix** [comment]
Hi! Thanks for your contribution! Have you solved this problem now? I'm also working on this problem, but I still have no solutions... I'm wondering, does your solution (superbase + nuxt content) works well now?

**Comment by sudomk** [comment]
I just thought about this conversation again and wrote the following simple middleware.

```
export default defineEventHandler((event) => {
    const url = event.req.url
    
    if(url.includes('/api/_content')) {
        return sendError(event, createError({
            statusCode: 401,
            statusMessage: 'Unauthorized'
        }))
})
```

Now when I make a request to `/api/_content/*` I get a 401. Even with `/api/_content/cache.json`. Or am I missing something?

**Reply by oripka** [reply]
Yes, of course. That seems to solve the open access to the content route.

**Comment by oripka** [comment]
I had a lot of performance issues with my solution above. It turned out that this part of the solution was the issue:

> I now solved this by overwriting pages/document-driven.vue from vue content, adding a middleware:
> 
> definePageMeta({
>     middleware: 'auth'
> })

Not sure how to protect the document driven pages without using an auth middleware and running into these performance issues

**Comment by sudomk** [comment]
A completely different solution could be to relocate your content to supabase storage and use their authentication.

**Comment by TheAnachronism** [comment]
Has anyone ever come up with a proper solution to this? Considering Nuxt and Nuxt Content have quite evolved in the past two years...

**Reply by oripka** [reply]
Yes — document-driven mode no longer exists, and the storage layer has been completely rewritten. Fortunately, the performance issues are gone as well.

In Nuxt Content v2, the concept of **collections** is used https://content.nuxt.com/docs/collections/define, and you can control access at the collection level by intercepting Nuxt Content API requests to a specific collection.

My current solution looks like something like this (`server/middleware/auth.ts`):

```ts
export default defineEventHandler(async (event) => {
  const url = event.node.req.url

  function getCollectionFromUrl(url: string) {
    const regex = /__nuxt_content\/([^/]+)/
    const match = url.match(regex)
    if (match && match[1]) return match[1]
    return null // or an empty string if preferred
  }

  if (url?.includes('/__nuxt_content/') && process.env.NODE_ENV !== 'prerender') {
    const collection = getCollectionFromUrl(url)
    // Check if the user has access to this collection...
  }
})
```

Here’s a related pull request, merged a couple of months ago, that also improves SSR handling:
https://github.com/nuxt/content/pull/3385

**Reply by TheAnachronism** [reply]
I've come up with a similar solution. The main thing from Nuxt content I need is the ability to write Markdown and turn it into a nice website, so the client-side feature of downloading the database isn't really a requirement. Furthermore, I've found that with a giant number of items in a collection, the database can get quite a bit, and in some experiments I had a client download 40 MB of one database dump. Because of this, I've moved all my querying to the server side, as most users will never browse a large percentage of a collection in a short time, making the client-side navigation with the database unnecessary.

The thing I've struggled with the most, tbh. is the way my deployment target, Cloudflare pages does things. Nuxt content also spits out a .sql version of the collection dumps that land in the root directory and get added to the exclude list in the `_routes.json` file. Because of this, and because Cloudflare pages considers everything not listed in the `include` part of the `_routes.json` file a static asset, these .sql files are publicly available too.
My solution for this is still not entirely complete. Right now, I just added a routeRule, that redirects access to these dump files to `/`, but there are certain caveats to this.

**Reply by oripka** [reply]
I didn't even realize that it created these .sql files (I also use nuxthub and cloudflare). I will start experimenting using your routeRules idea. Which caveats did you have with this solution?

I had a look at the code and it seems that nuxt content uses useStorage (https://github.com/nuxt/content/blob/main/src/runtime/presets/cloudflare/database-handler.ts) for the sql files. So it should be possible to put the sql dumps into a workers KV and remove the public files.

**Reply by oripka** [reply]
Ah I see, when configuring the redirects, I get errors when client-side navigating the restricted nuxt content but hard reload (ssr) seems to work. Did you find a workaround for this?

**Reply by TheAnachronism** [reply]
Well, first of all, when nuxt content detects that the stuff in the database doesn't align with the hash it has from the .SQL files, it makes those updates itself. I was running into timeouts and worker exhaustion on Cloudflare, due to the huge number of files and the sizes of them. Because of this, I eventually just made my own little migration script, that basically does this database migration that nuxt content would usually do on startup. Quite tedious, I know, but I just really wanted this to work with the current set up.
Furthermore, due to how Cloudflare pages does its things, I have to remove the `dump.*.sql` files from the `_routes.json`. I do this in my own nuxt module, with a `build:done` hook and checks for when the actual `_routes.json` file is generated. I haven't found a hook, yet that is executed _after_ nitro has completely built the entire `cloudflare-pages`, and something like `vite:compiled` can't be used, as in Nuxt this somehow _overrides_ existing compiled hooks, causing nitro to not even generate the `_routes.json` or `_redirects` files -.-

```ts
    nitro: {
        preset: "cloudflare-pages",
        prerender: {
            ignore: ["/"],
        },
    },
```

With this, I also disable the prerendering and generation of the .txt SQL dumps in the `__nuxt_content` directory, which would be downloaded by the client. One would think this would already be handled by a custom server middleware, but again Cloudflare comes in and messes this up, as those files are also added to the `exclude` section of the `_routes.json` file. Because of this, Cloudflare won't send requests for these files to the actual nuxt server, and just serve them statically, meaning the request never reaches the middleware to block it. With this, a bit overkill, prerender rule, those files aren't generated, and not added to `_routes.json`.
Finally, with the `_routes.json` having the SQL files removed, a simple `_redirects` rule like:
```
/dump.*	/	307
```
does the job. You can't set it to 404 or something, as it _has_ to be a redirect.
That's basically what I did, but I haven't tested yet, if Nuxt can still update the database itself, as it needs access to the SQL dumps. Not sure if all those rules could cause this to fail, but as I sort of do my own migration, I don't really have this problem.

**Reply by oripka** [reply]
I also see this in my `dist` folder `database/queries/content-database-001.sql` and is also downloadable but not in the excluded list of `_routes.json` maybe you have that, too?

**Reply by TheAnachronism** [reply]
In a normal build, my .SQL files are in the `exlcuded` list:
```json
{
  "version": 1,
  "include": [
    "/*"
  ],
  "exclude": [
    "/_nuxt/*",
    "/dump.posts.sql",
    "/favicon.ico",
    "/robots.txt",
    "/icons/archivebox.png"
  ]
}
```

I have this custom module here, that makes sure they are removed from that, so that Cloudflare doesn't serve them:
```ts
// modules/fix-routes.ts
import { defineNuxtModule } from "nuxt/kit";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export default defineNuxtModule({
    meta: {
        name: "fix-routes",
    },
    setup(_, nuxt) {
        const routesPath = join(nuxt.options.rootDir, "dist", "_routes.json");
        nuxt.hooks.afterEach((ctx) => {
            if (existsSync(routesPath) && ctx.name === "build:done") {
                const routes = JSON.parse(
                    readFileSync(routesPath, "utf-8"),
                ) as { exclude: string[] };

                const exclude = routes.exclude;
                if (Array.isArray(exclude)) {
                    for (let i = exclude.length - 1; i >= 0; i--) {
                        if (
                            typeof exclude[i] === "string" &&
                            exclude[i]!.endsWith(".sql")
                        ) {
                            exclude.splice(i, 1);
                        }
                    }
                }

                routes.exclude = exclude;

                writeFileSync(routesPath, JSON.stringify(routes, null, 2));
            }
        });
    },
});
```
It's still pretty hacky, as I make those checks after basically every hook and test for the hook name, but it works at the moment.
Weird that they don't show up on your side, you're using the `cloudflare-pages` nitro preset?

**Reply by oripka** [reply]
Thanks for sharing! Regarding your last point: I also see it appear on my side, but in addition I get a `database/queries/content-database-001.sql` file exposed publicly, which seems to come from NuxtHub.

I’m leaning toward opening a pull request, since all these workarounds feel hacky — and as you mentioned, they go against the original design philosophy of Nuxt Content. I also looked at Astro’s content implementation (since Nuxt Content v3 was partly inspired by Astro’s collections), but I didn’t find anything there related to securing access.

Here’s what I plan to explore:

* Secure access to the content databases (`database/queries/*`, `__nuxt_content/*`, `dump.*.sql`).

The main files involved seem to be:

```
src/presets/node.ts
src/module.ts
src/presets/nuxthub.ts
src/runtime/internal/api.ts
src/presets/cloudflare.ts
src/runtime/internal/database.server.ts
src/runtime/internal/dump.ts
src/utils/templates.ts
src/runtime/internal/database.client.ts
src/runtime/presets/cloudflare/database-handler.ts
```

## Ideas

| Option                                   | How it works                                                                                                              | Pros                                                                                 | Cons / Trade-offs                                                                      |
| ---------------------------------------- | ------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------- |
| **1. Nitro storage (v2-like)**           | Put dumps into Nitro’s internal storage; serve via API after auth                                                         | Simple; no extra infra                                                               | Bigger bundles; slower deploys/cold starts; goes against nuxt content v3’s goals                    |
| **2. R2/KV (private store)**             | Upload dumps to R2/KV; API serves/ signs URLs after auth                                                                  | Tiny function bundle; scales well                                                    | Requires extra infra + deploy wiring                                                   |
| **3. CF Access perimeter**               | Keep dumps static; protect path with Cloudflare Access                                                                    | Easiest if org-wide SSO ok                                                           | Coarse-grained; no per-user ACL                                                        |
| **4. **Encrypted dumps**** | Publish **encrypted** dumps (static/CDN). After auth, client gets a short-lived key, decrypts locally, then hydrates WASM | Works with static hosting/CDN; small bundle; no new infra; raw content isn’t exposed | Add build-time encryption + client decrypt; you must manage keys (short TTL, rotation) |

---

I lean toward **encryption (Option 4)** because it’s **simple**, **works with static public hosting/CDNs**, and adds **no external dependencies** (no R2/KV required). You keep v3’s architecture (thin server bundle, client-side WASM), while making the exposed artifacts useless without a key.

**Reply by oripka** [reply]
Pull request opened: https://github.com/nuxt/content/pull/3510.
It has a lot of changes so I doubt it will land anytime soon...

**Reply by BenjaminOddou** [reply]
Hello @oripka, your solution seems very promising ! Did you notice any performance degradation when decrypting large amount of data ? Do you think that your modifications to nuxt content could be added in a separate module while waiting the merge of your pr (if you have some recommandation on how to handle it). 


Also, do you think it would be possible to have a per page decryption ? If i understand correctly, you are encryping /decrypting the full collections but if i'd like to give the access to some part of the collection it's seems not possible. 

Many thanks in advance and thanks for the great work. 👏

**Reply by oripka** [reply]
Hello @BenjaminOddou ,

That’s a good idea. I’ll go ahead and publish a module on npm — I was hesitant at first since I’d prefer not to maintain a fork.

Regarding decryption, I haven’t tested with very large datasets yet, but here’s how it currently works:
	•	Decryption only runs during the initial hydration of a collection when its checksum is missing or has changed.
	•	The process (AES-GCM decrypt + gzip decompress) happens once per collection load, not per query.
	•	Keys and dumps are cached in localStorage, so reloads usually avoid extra server calls and only rehydrate the database in memory.
	•	After hydration, all queries run directly against an in-memory SQLite database without triggering decryption again.

The performance cost is therefore limited to this one-time hydration step, which scales with dump size. On low-powered devices, hydration may feel slower, but once complete, navigation and queries are as fast as with plain text.

As for per-page decryption: in theory it could be done, but with Nuxt Content the API is structured around collections (they’re part of the query parameters). It feels natural to treat a collection as the security boundary (that you can also incept in middlewares, see the example in the docs) rather than working against that model. If you need finer separation, you can always create additional collections.

**Reply by oripka** [reply]
Here is a package that is based upon the fork: https://www.npmjs.com/package/@ripka/content

To test drive one needs to adjust he `package.json` to override `@nuxt/content` 

```
...
  "pnpm": {
    "onlyBuiltDependencies": [
      "better-sqlite3"
    ],
    "overrides": {
      "vite": "npm:rolldown-vite@latest",
      "@nuxt/content": "npm:@ripka/content@latest"
    }
  },
...
```

Run `pnpm install`

And enable the encryption via the `nuxt.config.ts`


```
export default defineNuxtConfig({
---
  content: {
    encryption: {
      enabled: true
    },
...
```

**Reply by BenjaminOddou** [reply]
> As for per-page decryption: in theory it could be done, but with Nuxt Content the API is structured around collections (they’re part of the query parameters). It feels natural to treat a collection as the security boundary (that you can also incept in middlewares, see the example in the docs) rather than working against that model. If you need finer separation, you can always create additional collections.

That's what I thought. In my app, each user can is assigned to groups that are then used in the frontmatter of pages to control the visibility. I would have prefer decrypting pages instead of full collection but indeed it doesn't seems to be aligned with the nuxt content philosophy.

Do you know what are the routes that lead to the database / dumps in production ?

**Reply by oripka** [reply]
If you are okay with losing the ability to have offline access and downloading the dump to the client side, you can always take the approach of @TheAnachronism, block the dump download, and only allow access via API on the server-side. You lose some benefits, but then you can do whatever in the middleware to check parameters.
